### ISSUE TITLE
Request the download link of Baidu online disk

### ISSUE THREAD

**Issue by TianX1nYu** [luyvlei/BiSMAPs#3]
Thank you for your excellent work. Can you provide the download link of Stylized Datasets on Baidu's online disk?

**Comment by luyvlei** [comment]
No problem, but it will take some time.

**Comment by TianX1nYu** [comment]
Sorry to bother you again, I have downloaded your synthia_ Stylized.tar.gz file, but there is a problem. Is there a problem with 000000. png? It occupies more than 10GB and cannot be opened

**Comment by luyvlei** [comment]
I checked that this should be caused by the wrong operation when creating the tar.gz file. I have re-upload a correct version 'synthia_stylized.zip'.  Due to the upload size limit of Baidu Netdisk, I will upload to Aliyun Drive instead.

**Comment by TianX1nYu** [comment]
Sorry to bother you again. By observing and loading the synthia warmup weights you provided, I found that the following weights will not be loaded into the model, but they do not seem to affect the results. What are these weights?
![image](https://user-images.githubusercontent.com/88664189/191028396-bfbe197d-a6e0-450d-a8e5-7d6e84d0f051.png)

**Comment by luyvlei** [comment]
This part of the module is not used, and the parameters are randomly initialized. So just ignore it.

**Comment by luoxin13** [comment]
Thanks for sharing stylized GTA5 and SYNTHIA images! I am wondering if it is convenient for you to share the codes for stylization. So that we can apply the proposed methods for other types of synthetic datasets. :)

**Comment by luyvlei** [comment]
Sure, later I will share the code.---- Replied Message ----FromXin ***@***.***>Date09/28/2022 09:16 ***@***.***> ***@***.******@***.***>SubjectRe: [luyvlei/BiSMAPs] Request the download link of Baidu online disk (Issue #3)
Thanks for sharing stylized GTA5 and SYNTHIA images! I am wondering if it is convenient for you to share the codes for stylization. So that we can apply the proposed methods for other types of synthetic datasets. :)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/luyvlei/BiSMAPs/issues/3#issuecomment-1260268718",
"url": "https://github.com/luyvlei/BiSMAPs/issues/3#issuecomment-1260268718",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by luyvlei** [comment]
> Thanks for sharing stylized GTA5 and SYNTHIA images! I am wondering if it is convenient for you to share the codes for stylization. So that we can apply the proposed methods for other types of synthetic datasets. :)

Sure, I will share the code later.

**Comment by luyvlei** [comment]
> Thanks for sharing stylized GTA5 and SYNTHIA images! I am wondering if it is convenient for you to share the codes for stylization. So that we can apply the proposed methods for other types of synthetic datasets. :)

You can use the following code, the version of scikit-image is 0.18.2. 

This script is reimplement according to https://arxiv.org/abs/2103.13041. They integrated this function into the dataloader to get more diverse stylized images, but I didn't achieve a significant gain in my experiments, so I used an offline method to generate images to reduce training overhead.

```
import argparse
import torch
import cv2
import os
import numpy as np
from skimage.exposure import match_histograms
from tqdm import tqdm
from multiprocessing import Pool
import time

def inf_sampler(list):
    length = len(list)
    cur = 0
    while True:
        cur += 1
        cur %= length
        yield list[cur]

def cal_gamma_numpy(source_img_L, target_img_L, min_iter, max_iter, lr, beta, delta):
    """calculate gamma use numpy
    """
    target_img_L = target_img_L[0:-1:16,0:-1:16]
    source_img_L = source_img_L[0:-1:16,0:-1:16]
    gamma = 1
    target_img_Lmean = target_img_L.mean()
    source_img_Llog = np.log(np.clip(source_img_L,1e-3,1.0))

    for i in range(max_iter):
        # cal grad
        # target function is : target = ((source_img_L**gamma).mean() - target_img_Lmean)**2 + beta * (gamma-1)**2
        # grad function is : grad =  2 * ((source_img_L**gamma).mean() - target_img_Lmean) * (np.log(source_img_L) * source_img_L**gamma).mean() + 2 * beta * (gamma-1)
        L_gamma = (source_img_L**gamma)
        L_gamma_mean = L_gamma.mean()
        L_dis = (L_gamma_mean - target_img_Lmean)
        
        # grad function
        grad_gamma = 2 * L_dis * (source_img_Llog * L_gamma).mean() + beta * 2 * (gamma-1)
        gamma -= grad_gamma * lr
        if i > min_iter and abs(grad_gamma) < delta:
            break
    return gamma

def main_worker(idx, source, target, cal_gamma_fn, args):
    source_img = cv2.cvtColor(cv2.imread(source,1), cv2.COLOR_BGR2Lab).astype(np.float64)
    source_img /= 255.
    target_img = cv2.cvtColor(cv2.imread(target,1), cv2.COLOR_BGR2Lab).astype(np.float64)
    target_img /= 255.

    # cal gamma
    source_img_L, target_img_L = source_img[...,0], target_img[...,0]
    gamma = cal_gamma_fn(source_img_L, target_img_L, args.min_iter, args.max_iter, args.lr, args.beta, args.delta_min)

    # hist match a,b
    source_img[:,:,1:3] = match_histograms(source_img[:,:,1:3], target_img[:, :, 1:3], multichannel=True)
    source_img[:,:,0] = source_img[:,:,0]**gamma

    source_transfered = (source_img*255).astype(np.uint8)
    source_transfered[source_transfered>=255]=255
    source_transfered = cv2.cvtColor(source_transfered, cv2.COLOR_Lab2BGR)
    cv2.imwrite(os.path.join(args.output_path, source.split("/")[-1]), source_transfered)
    print("\r {}/{}".format(idx, args.length), end="")


def main(args):
    if not os.path.exists(args.output_path):
        os.mkdir(args.output_path)
    with open(os.path.join(args.output_path, "args.txt"), "w") as f:
        f.write(str(args.__dict__))

    source_list = [os.path.join(args.source_path, i) for i in os.listdir(args.source_path)]
    target_list = [os.path.join(args.target_path, i, j) for i in os.listdir(args.target_path) for j in os.listdir(os.path.join(args.target_path,i))]
    source_list.sort()
    args.length = len(source_list)
    
    target_sampler = inf_sampler(target_list)

    cal_gamma_fn = cal_gamma_numpy

    if  args.workers>0:
        mp_pool = Pool(args.workers)
        for idx, source in enumerate(source_list):
            target = next(target_sampler)
            res = mp_pool.apply_async(main_worker, (idx, source, target, cal_gamma_fn, args))
        mp_pool.close()
        mp_pool.join()
    else:
        for idx, source in tqdm(enumerate(source_list)):
            target = next(target_sampler)
            main_worker(idx, source, target, cal_gamma_fn, args)


if __name__=='__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--beta', type=float, default=0.01)
    parser.add_argument('--source_path', type=str, default="./dataset/GTA5/images/")
    parser.add_argument('--target_path', type=str, default="./dataset/CityScape/leftImg8bit/train/")
    parser.add_argument('--output_path', type=str, default="gta_stylized")
    parser.add_argument('--delta_min', type=int, default=3e-3)
    parser.add_argument('--min_iter', type=int, default=5)
    parser.add_argument('--max_iter', type=int, default=50)
    parser.add_argument('--lr', type=float, default=1.0)
    parser.add_argument('--workers', type=int, default=4)
    args = parser.parse_args()
    main(args)
```

**Comment by luoxin13** [comment]
> 

Thanks so much for sharing these codes!

**Comment by TianX1nYu** [comment]
![image](https://user-images.githubusercontent.com/88664189/194788214-45212b4e-5c40-46a2-8642-8d83ddf66f61.png)
Sorry to bother you again，Your visualization results in Figure 6 are very beautiful. Can you share the code for drawing feature distribution map with T-SNE? Thank you very much

**Comment by luyvlei** [comment]
> Sorry to bother you again，Your visualization results in Figure 6 are very beautiful. Can you share the code for drawing feature distribution map with T-SNE? Thank you very much

Only the scatter diagrams are generated by python. The layout of other formats is completed using powerpoint. I can give you the code of the t-sne scatterplot later if needed.

**Comment by Jack-bo1220** [comment]
> 

Hi, could your the code about  the t-sne scatterplot. Thank you!
